win32ss/user/user32/windows/input.c

```diff
@@ -330,6 +355,187 @@ LoadKeyboardLayoutW(LPCWSTR pwszKLID,
+#define FE_KOREAN               (1 << 3)
+
+// Sets the far-east flags
+// Win: SetFeKeyboardFlags
+VOID FASTCALL IntSetFeKeyboardFlags(LANGID LangID, PBYTE pbFlags)
+{
+    switch (LangID)
+    {
+        case MAKELANGID(LANG_JAPANESE, SUBLANG_DEFAULT):
+            *pbFlags |= FE_JAPANESE;
+            break;
+
+        case MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_TRADITIONAL):
+        case MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_HONGKONG):
+            *pbFlags |= FE_CHINESE_TRADITIONAL;
+            break;
+
+        case MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_SIMPLIFIED):
+        case MAKELANGID(LANG_CHINESE, SUBLANG_CHINESE_SINGAPORE):
+            *pbFlags |= FE_CHINESE_SIMPLIFIED;
+            break;
+
+        case MAKELANGID(LANG_KOREAN, SUBLANG_KOREAN):
+            *pbFlags |= FE_KOREAN;
+            break;
+
+        default:
+            break;
+    }
+}
+
+BOOL APIENTRY
+CliImmSetHotKeyWorker(DWORD dwHotKeyId, UINT uModifiers, UINT uVirtualKey, HKL hKL, DWORD dwAction)
+{
+    if (dwAction == SETIMEHOTKEY_ADD)
+    {
+        if (IME_HOTKEY_DSWITCH_FIRST <= dwHotKeyId && dwHotKeyId <= IME_HOTKEY_DSWITCH_LAST)
+        {
+            if (!hKL)
+                goto Failure;
+        }
+        else
+        {
+            if (hKL)
+                goto Failure;
+
+            if (IME_KHOTKEY_SHAPE_TOGGLE <= dwHotKeyId &&
+                dwHotKeyId < IME_THOTKEY_IME_NONIME_TOGGLE)
+            {
+                goto Failure;
+            }
+        }
+
+#define MOD_ALL_MODS (MOD_ALT | MOD_CONTROL | MOD_SHIFT | MOD_WIN)
+        if ((uModifiers & MOD_ALL_MODS) && !(uModifiers & (MOD_LEFT | MOD_RIGHT)))
+        {
+            goto Failure;
+        }
+#undef MOD_ALL_MODS
+    }
+
+    return NtUserSetImeHotKey(dwHotKeyId, uModifiers, uVirtualKey, hKL, dwAction);
+
+Failure:
+    SetLastError(ERROR_INVALID_PARAMETER);
+    return FALSE;
+}
+
+VOID APIENTRY CliSetDefaultImeHotKeys(PMYHOTKEYENTRY pEntries, UINT nCount, BOOL bCheck)
+{
+    UINT uVirtualKey, uModifiers;
+    HKL hKL;
+
+    while (nCount-- > 0)
+    {
+        if (!bCheck || !NtUserGetImeHotKey(pEntries->dwHotKeyId, &uModifiers, &uVirtualKey, &hKL))
+        {
+            CliImmSetHotKeyWorker(pEntries->dwHotKeyId,
+                                  pEntries->uModifiers,
+                                  pEntries->uVirtualKey,
+                                  pEntries->hKL,
+                                  SETIMEHOTKEY_ADD);
+        }
+        ++pEntries;
+    }
+}
+
+VOID APIENTRY CliGetPreloadKeyboardLayouts(BYTE* pbFlags)
+{
+    WCHAR szValue[9], szKeyName[4];
+    UNICODE_STRING ValueString;
+    DWORD dwKL, ret;
+    UINT iNumber;
+
+    for (iNumber = 1; iNumber < 1000; ++iNumber)
+    {
+        StringCchPrintfW(szKeyName, _countof(szKeyName), L"%d", iNumber);
+        ret = GetPrivateProfileStringW(L"Preload", szKeyName, L"", szValue, _countof(szValue),
+                                       L"keyboardlayout.ini");
```

Comment:
lol, why?

Reply:
What?

Reply:
This code will check the `[Preload]` settings of File `keyboardlayout.ini`

Reply:
yes of course i understand what the code reads. My reaction was more about why it uses an INI file for that instead of what's being usually done, i.e. reading the settings from the registry. And there is literally zero explanation for that choice in the code...

Reply:
Added a comment.

Reply:
Should we use the `Reg*` functions instead of the profile function at here?

Reply:
Literally the very same thing as what is done in user32's LoadKeyboardLayoutW function...